### ISSUE TITLE
Possible issue with older versions of ggplot2

### ISSUE THREAD

**Issue by jeff-hughes** [topepo/caret#571]
I recently asked and answered a question on Stack Overflow about an error I ran into with the _caret_ package, possibly as a result of the _lazyeval_ package. I haven't been able to specifically verify that my explanation for the issue is correct, but I did find a solution that may help to track down the reason, at least. This error arose when I attempted to train any model whatsoever, but I do provide a reproducible example, as well as my session info before and after, on the Stack Overflow question here:

http://stackoverflow.com/questions/41558992/r-caret-package-running-into-compile-environment-error/41579786

My best guess is that perhaps a function from _lazyeval_ is being used in the _caret_ package without specifying it as a dependency, and thus older versions of _ggplot2_ run into a problem. Caret loads _ggplot2_, and the newest version of _ggplot2_ loads _lazyeval_, but older versions do not.

I could be off base, as it was a very strange error I had to track down, but I think you may be able to resolve this either by specifying lazyeval as a dependency, or specifying ggplot2 >= 2.2.

**Comment by jeff-hughes** [comment]
Apparently I spoke too soon. I swear I had gotten this working, but upon trying this an hour later, it is back to giving me the same error message. The _compiler_ package is not getting loaded for some reason, and I have no idea why. This is despite _lazyeval_ getting loaded, so that explanation appears to be off-base.

**Comment by topepo** [comment]
I'm not sure. I can't reproduce the issue with the same package versions (but on OS X): 
```r
> sessionInfo()
R version 3.3.2 (2016-10-31)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Sierra 10.12.2

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] caret_6.0-73    ggplot2_2.2.1   lattice_0.20-34

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.8        magrittr_1.5       splines_3.3.2      MASS_7.3-45       
 [5] munsell_0.4.3      colorspace_1.3-1   foreach_1.4.3      minqa_1.2.4       
 [9] stringr_1.1.0      car_2.1-4          plyr_1.8.4         tools_3.3.2       
[13] parallel_3.3.2     nnet_7.3-12        pbkrtest_0.4-6     grid_3.3.2        
[17] gtable_0.2.0       nlme_3.1-128       mgcv_1.8-15        quantreg_5.29     
[21] MatrixModels_0.4-1 iterators_1.0.8    lme4_1.1-12        lazyeval_0.2.0    
[25] assertthat_0.1     tibble_1.2         Matrix_1.2-7.1     nloptr_1.0.4      
[29] reshape2_1.4.2     ModelMetrics_1.1.0 codetools_0.2-15   stringi_1.1.2     
[33] scales_0.4.1       stats4_3.3.2       SparseM_1.74    
```
I can try on a windows machine when I'm back home. 

Did you update R to a significantly different version (e.g. major) without reinstalling packages? Have you tried reinstalling `caret` too?

**Comment by jeff-hughes** [comment]
I upgraded from R 3.2 to 3.3 somewhat recently, but most of my packages are up to date or close to it. I did try reinstalling caret, and reinstalling R as well in case the compiler package somehow got corrupted. I haven't come across this issue in any other context, and it works on my laptop but not on my desktop (running the same version numbers and both Win10).

**Comment by jeff-hughes** [comment]
I went through and updated each of these packages one by one to try to isolate which one would resolve the problem. After updating all of them, it still gives me the same error. Here's my new session info:

    > sessionInfo()
    R version 3.3.2 (2016-10-31)
    Platform: x86_64-w64-mingw32/x64 (64-bit)
    Running under: Windows >= 8 x64 (build 9200)
    
    locale:
    [1] LC_COLLATE=English_United States.1252  LC_CTYPE=English_United States.1252    LC_MONETARY=English_United States.1252
    [4] LC_NUMERIC=C                           LC_TIME=English_United States.1252    
    
    attached base packages:
    [1] stats     graphics  grDevices utils     datasets  methods   base     
    
    other attached packages:
    [1] caret_6.0-73    ggplot2_2.2.1   lattice_0.20-34
    
    loaded via a namespace (and not attached):
     [1] Rcpp_0.12.8        magrittr_1.5       splines_3.3.2      MASS_7.3-45        munsell_0.4.3      colorspace_1.3-2  
     [7] foreach_1.4.3      minqa_1.2.4        stringr_1.1.0      car_2.1-4          plyr_1.8.4         tools_3.3.2       
    [13] parallel_3.3.2     nnet_7.3-12        pbkrtest_0.4-6     grid_3.3.2         gtable_0.2.0       nlme_3.1-128      
    [19] mgcv_1.8-16        quantreg_5.29      e1071_1.6-7        class_7.3-14       MatrixModels_0.4-1 iterators_1.0.8   
    [25] lme4_1.1-12        lazyeval_0.2.0     assertthat_0.1     tibble_1.2         Matrix_1.2-7.1     nloptr_1.0.4      
    [31] reshape2_1.4.2     ModelMetrics_1.1.0 codetools_0.2-15   stringi_1.1.2      scales_0.4.1       stats4_3.3.2      
    [37] SparseM_1.74  

The only differences I can see (apart from being on Windows rather than Mac) is that my _mgcv_ package was more up-to-date than yours. And now, after updating, my _colorspace_ package is more up-to-date than yours. Neither of those really gives much of a hint at what the problem might be.

**Comment by jeff-hughes** [comment]
I think I have **finally** figured out the issue. I had set my library location to a global folder so I wouldn't have to move all my packages from the "3.2" folder to the "3.3" folder, etc. every time I updated R. My solution worked, but had the side effect of removing the default library location in Program Files (the one that includes the base packages).

That seemed to work completely fine for everything I did, so I never noticed an issue until I started using caret. But for some reason, apparently the compiler package (used by some dependency of caret) doesn't work properly with that setup, and needs an explicit reference to the Program Files library location in .libPaths(). After changing things back to the default, everything seems to work fine.

So I'd rather just deal with the slight inconvenience of moving packages when updating R, in order to be able to use the massively useful caret package. As such, I'll close this issue given that it's completely idiosyncratic and nothing to do with caret in particular. Final note: "Don't try to be smarter than the R maintainers. Just put up with the damned inconvenience and move on."
